relayer/network_utils.go

```diff
@@ -0,0 +1,139 @@
+) error {
+	subnetID := sourceBlockchain.GetSubnetID()
+	connectedValidators, err := network.ConnectToCanonicalValidators(subnetID)
+	if err != nil {
+		logger.Error(
+			"Failed to connect to canonical validators",
+			zap.String("subnetID", subnetID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
+	for _, destination := range sourceBlockchain.SupportedDestinations {
+		blockchainID := destination.GetBlockchainID()
+		if ok, quorum, err := checkForSufficientConnectedStake(logger, cfg, connectedValidators, blockchainID); !ok {
+			logger.Error(
+				"Failed to connect to a threshold of stake",
+				zap.String("destinationBlockchainID", blockchainID.String()),
+				zap.Uint64("connectedWeight", connectedValidators.ConnectedWeight),
+				zap.Uint64("totalValidatorWeight", connectedValidators.TotalValidatorWeight),
+				zap.Any("warpQuorum", quorum),
+			)
+			return err
+		}
+	}
+	return nil
+}
+
+// Connect to the validators of the destination blockchains. Verify that we have connected
+// to a threshold of stake for each blockchain.
+func connectToPrimaryNetworkPeers(
+	logger logging.Logger,
+	network *peers.AppRequestNetwork,
+	cfg *config.Config,
+	sourceBlockchain *config.SourceBlockchain,
+) error {
+	for _, destination := range sourceBlockchain.SupportedDestinations {
+		blockchainID := destination.GetBlockchainID()
+		subnetID := cfg.GetSubnetID(blockchainID)
+		connectedValidators, err := network.ConnectToCanonicalValidators(subnetID)
+		if err != nil {
+			logger.Error(
+				"Failed to connect to canonical validators",
+				zap.String("subnetID", subnetID.String()),
+				zap.Error(err),
+			)
+			return err
+		}
+
+		if ok, quorum, err := checkForSufficientConnectedStake(logger, cfg, connectedValidators, blockchainID); !ok {
+			logger.Error(
+				"Failed to connect to a threshold of stake",
+				zap.String("destinationBlockchainID", blockchainID.String()),
+				zap.Uint64("connectedWeight", connectedValidators.ConnectedWeight),
+				zap.Uint64("totalValidatorWeight", connectedValidators.TotalValidatorWeight),
+				zap.Any("warpQuorum", quorum),
+			)
+			return err
+		}
+	}
+	return nil
+}
+
+// Fetch the warp quorum from the config and check if the connected stake exceeds the threshold
+func checkForSufficientConnectedStake(
+	logger logging.Logger,
+	cfg *config.Config,
+	connectedValidators *peers.ConnectedCanonicalValidators,
+	destinationBlockchainID ids.ID,
+) (bool, *config.WarpQuorum, error) {
+	quorum, err := cfg.GetWarpQuorum(destinationBlockchainID)
+	if err != nil {
+		logger.Error(
+			"Failed to get warp quorum from config",
+			zap.String("destinationBlockchainID", destinationBlockchainID.String()),
+			zap.Error(err),
+		)
+		return false, nil, err
+	}
+	return utils.CheckStakeWeightExceedsThreshold(
+		big.NewInt(0).SetUint64(connectedValidators.ConnectedWeight),
+		connectedValidators.TotalValidatorWeight,
+		quorum.QuorumNumerator,
+		quorum.QuorumDenominator,
+	), &quorum, nil
+}
```

Comment:
Going along with my above comment - I think this would actually be better served _not_ being a function - the tri-state return is kind of confusing. I think we should just inline this in the calling function. 

If we wanted, we could make a helper that takes `connectedValidators` and a `quorum` and returns a bool to do the calculation on lines 133-138

Reply:
I like the suggestion to break up the quorum check and the stake weight check into separate functions.